lib/pbench/server/api/resources/__init__.py

```diff
@@ -789,7 +791,7 @@ def _dispatch(
         # We don't accept or process a request payload for GET, or if no
         # parameter schema is defined
         if not self.schema or method == self._get:
-            return method(uri_encoded, request)
+            return method(uri_parameters, request)
 
         try:
             json_data = request.get_json()
```

Comment:
So in `_dispatch` the `json_data` immediately goes through the validation (self.schema.validate(json_data)) and if it is `None` or empty `{}`, we raise `InvalidRequestPayload`. Check line 535 of `.../api/resources/__init__.py`. So I dont think it would be a cause of concern, since if the payload is missing and `uri_parameter` is also empty `{}` it will be a `BAD_REQUEST`, and if the payload is missing but  `uri_parameter` is not empty `{"x":"y"}` then it is possible that is might raise `BAD_REQUEST` because of missing parameter in the payload (if there is any required parameter in our class definition, and if not we will proceed with the request).

Reply:
What if the request includes a body which has a MIME type of `application/octet-stream`?  My understanding is, even if the body is valid JSON, the call to `request.get_json()` will return `None`.  We won't detect this until the call to `self.schema.validate()` which happens some time later; and, if there are URL parameters which are present, we won't reject the request unless the schema specifies required parameters in the body.  

We should probably specify`force=true` in the `request.get_json()` call if we're not going to reject the request based on the payload MIME type.

My instinct is that `request.get_json()` should return `None` (or raise an exception) on error; otherwise, it should return `{}` or the contents of the payload.  So, I feel like we should be catching and reporting the error (i.e., the `None`) case before proceeding.  I think we should be treating `uri_parameters` the same way.

And, having done that, by the time we get to the following `try` block, we can just add the contents of `uri_parameters` to `json_data` without any conditional and proceed on our way.

Reply:
> What if the request includes a body which has a MIME type of `application/octet-stream`?

Yeah as you said we can probably specify `force=true` in the request.get_json() to restrict to `application/json`

But the thing about the `uri_parameter` is that it is either going to be something or the user is going to get `NOT_FOUND` so `uri_parameter` can either be `{}` or `{"x":"y"}` but never `None`. Even with the current code we do check the `request.get_json()` for None and unless all the parameters in the json are validated according to the class schema definition we don't proceed. Its just that we do consume the json as it is and then validate it by our implementation and not Flask's framework.

Reply:
> Yeah as you said we can probably specify `force=true` in the `request.get_json()` to restrict to `application/json`

By default, `request.get_json()` [will return `None`](https://flask.palletsprojects.com/en/latest/api/#flask.Request.get_json) if the body MIME type is not `application/json` (i.e., `force` is `false` by default); we would specify `force=true` if we wanted to be able to ignore the MIME type.

Being able to depend on the `uri_parameter` being a `dict` in our code is handy.  I would like to have the same assurance for `json_data`, not only for convenience but also because I think the cases in which it is not a `dict` represent errors which we want to respond to and not ignore or cover over.

Reply:
Yeah, I've thought several times about enforcing `Content-Type` and even `Accepts` since we *return* JSON. It'd actually be cool if Flask would transparently handle `dict` <-> (`application/json` | `application|xml`), but I don't think it does and it's not really a big deal. But we probably shouldn't be just silently assuming JSON everywhere, and I never really got motivated to try to do anything about it.

Reply:
Yeah I agree, but at this time its not catastrophic if the `get_json` returns `None` because if the conditional statement on line 813, but yes we should be enforcing `Content-Type` to be `application/json`. I will work on it separately since I think it might need little bit experimentation